File: Programação Funcional/Coleções/Scala/Listas com Recursão/15.scala
object Main {
  //Função que retorna true se todos os elementos da lista forem menores ou iguais ao limite, e false caso contrário.
  def todosMenoresOuIguais(lista: List[Int], limite: Int): Boolean = lista match {
    case Nil => true
    case cabeca :: cauda =>
      if (cabeca <= limite) todosMenoresOuIguais(cauda, limite)
      else false
  }

  // Função principal
  def main(args: Array[String]): Unit = {
    val minhaLista = List(1, 2, 3, 4, 5)
    println(todosMenoresOuIguais(minhaLista, 5)) // Saída: true
    println(todosMenoresOuIguais(minhaLista, 4)) // Saída: false	
  }
}

File: Programação Lógica/Prolog/Listas/15.pl
soma_n(_, 0, 0).  % Caso base
soma_n([Cabeca | Cauda], N, Soma) :-
    N > 0, 
    N1 is N - 1,
    soma_n(Cauda, N1, SomaCauda),
    Soma is Cabeca + SomaCauda.
